Добавить валидации на поле title у счета #20

AllaVrublevskaja opened the pull request.

AllaVrublevskaja committed:
openapi.yaml - изменена длина поля title со 100 на 50.
OpenApiReader.java - добавлен класс для чтения длины поля title из openapi.yaml.
Добавлена зависимость в build.gradle.kts в пакет componentTest для чтения
openapi.yaml.
Создан тестовый json-файл create-title-validate-length.json для тестирования длины поля title.
AccountJson.java - добавлен путь к файлу create-title-validate-length.json.
AccountsComponentTest.java - добавлены два теста для валидации длины поля title.

AllaVrublevskaja committed:
AccountsComponentTest.java - изменен тест createAccount_WhenTitleExceedsMaxLengthFromOpenApi_ShouldFail в соответствии с требованиями задания.

AllaVrublevskaja committed:
Merge branch 'master' into Добавить_валидации_на_поле_title_у_счета

# Conflicts:
#	component-test/src/test/java/AccountJson.java
#	component-test/src/test/java/AccountsComponentTest.java

Defenx commented on component-test/src/test/java/OpenApiReader.java:
```diff
@@ -0,0 +1,58 @@
+        return titleMaxLength;
+    }
+
+    private static int readTitleMaxLengthFromOpenApi() {
```
очень странный метод, для чего он нужен?

AllaVrublevskaja replied:
Этот метод проходит по узлам дерева openapi.yaml, находит поле tiltle и его длину.
Мне показался этот класс очень интересным и нужным. Он позволяет провести валидацию любого поля на соответствие схеме. Получение длины поля из openapi.yaml c помощью класса OpenApiReader и этого метода дает возможность исключить рефакторинг тестов если длина поля опять изменится. А это повышает устойчивость тестов.

Defenx commented on component-test/build.gradle.kts:
```diff
@@ -22,6 +22,7 @@ dependencies {
     testImplementation(libs.spring.boot.testcontainers)
     testImplementation(libs.testcontainers.junit)
     testImplementation(libs.testcontainers.postgresql)
+    testImplementation("com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.15.2")
```
думаю, что когда переделаем тесты по стандартному, как уже сделано для другой логики, эта зависимость станет ненужна

AllaVrublevskaja replied:
Эта зависимость нужна для создания new YAMLFactory() в class OpenApiReader. Если этот класс не нужен, то не нужна и она.

Defenx commented on app/src/resources/static/swagger/openapi.yaml:
```diff
@@ -267,7 +267,7 @@ components:
         title:
           type: string
-          maxLength: 100
+          maxLength: 50
```
давай это ограничение впринципе уберем, мы установим ограничение только на создание счета
(вообще ограничение создаем для того чтобы красиво получилось отобразить на фронтенде название счета)

AllaVrublevskaja replied:
оставить только title:
          type: string
а maxLength: 50 - удалить?

AndreyTerex replied:
> оставить только title: type: string а maxLength: 50 - удалить?

Тут валера указывает на место в коде, где в схеме openApi мы указывает ограничение на responseDTO(270 строчка), ограничение на requestDTO нужно оставить(242 строчка).
Так что да, maxLength можно удалить на 270 строке

Defenx commented on app/src/resources/static/swagger/openapi.yaml:
```diff
@@ -239,7 +239,7 @@ components:
           type: string
           nullable: true
-          maxLength: 100
+          maxLength: 50
```
да, все верно

Defenx commented on component-test/src/test/java/AccountsComponentTest.java:
```diff
@@ -272,4 +273,41 @@ void createWithoutAuth401() throws Exception {
+                .andReturn().getResponse();
+
+        UUID id = TestUtils.extractIdFromLocation(response);
+        Account account = accountRepository.findById(id).orElseThrow();
```
да, это договоренность на проекте, в данный момент Андрей сделал задачу на исправление, если где-то в тестах можно заменить прямой поход в бд на использование API https://github.com/Defenx/cash-savings-project/pull/36

AllaVrublevskaja replied:
Хорошо. Посмотрю.

Defenx commented on component-test/src/test/resources/json/accounts/create-title-validate-length.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "title": "%s",
```
идея интересная, но я думаю пока, что можно просто конкретное значение проставить

Если заметим что в коде будут появлятся места где дублируем код или файлы тестовые, я заведу задачку на рефакторинг

AllaVrublevskaja replied:
эта запись позволяет построить значение title в соответствии с длиной title из схемы

AllaVrublevskaja commented on component-test/src/test/java/OpenApiReader.java:
```diff
@@ -0,0 +1,58 @@
+
+import java.io.InputStream;
+
+public class OpenApiReader {
```
Этот класс нужен для получения информации о любом поле из схемы, а это самая точная информация. Если длина поля title опять изменится, то не нужно будет вносить изменения в тесты

AndreyTerex replied:
если схема openapi измениться то возможно придётся и этот класс переделывать

AllaVrublevskaja replied:
да, но это будет одно место, которое надо будет менять

AndreyTerex replied:
Можно сделать константу с числом, где будет лежать максимальная длинна строки и метод который генерит строку с нужным кол-вом символов.
 Можно будет менять только значение этой константы, аналогично 1 место, но код гораздо проще будет

AndreyTerex replied:
нам не нужно читать openApi схему, нам нужно лишь проверить, что выпадет ошибка при нарушении constraints, этот класс это лишнее усложнение которое лишь добавит проблем в дальнейшей поддержке кода

AllaVrublevskaja commented on component-test/src/test/java/OpenApiReader.java:
```diff
@@ -0,0 +1,58 @@
+import java.io.InputStream;
+
+public class OpenApiReader {
+    private static final String OPENAPI_FILE = "static\\swagger\\openapi.yaml";
```
"static\swagger\openapi.yaml" - такой путь у меня дает ошибку(ели я не экранирую слеш).

AllaVrublevskaja commented on component-test/src/test/java/AccountsComponentTest.java:
```diff
@@ -272,4 +273,41 @@ void createWithoutAuth401() throws Exception {
+    @DisplayName("POST: title максимальной длины из OpenAPI -> 201")
+    void createAccount_WhenTitleMaxLengthFromOpenApi_ShouldSucceed() throws Exception {
+        int maxLength = OpenApiReader.getTitleMaxLength();
+        String validTitle = OpenApiReader.generateTitleOfLength(maxLength);
```
Если удалить OpenReader, то да. Но если изменится по какой-то причине длина title, то нужно будет вносить изменения в тест.

AllaVrublevskaja commented on component-test/src/test/java/AccountsComponentTest.java:
```diff
@@ -272,4 +273,41 @@ void createWithoutAuth401() throws Exception {
+                        .with(httpBasic(ACCOUNT_TEST_EMAIL, TEST_PASSWORD))
+                        .contentType(MediaType.APPLICATION_JSON_VALUE)
+                        .accept(MediaType.APPLICATION_PROBLEM_JSON_VALUE)
+                        .content(json))
```
Ok.

AllaVrublevskaja committed:
Merge branch 'master' into Добавить_валидации_на_поле_title_у_счета

# Conflicts:
#	component-test/src/test/java/AccountsComponentTest.java

AllaVrublevskaja committed:
OpenApiReader.java - Класс удален
build.gradle.kts - Удалена зависимость testImplementation("com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.15.2")

AllaVrublevskaja committed:
AccountsComponentTest.java - изменены тесты на соответствие длины title схеме
AccountJson.java - добавлена константа CREATE_TITLE_INVALIDATE_LENGTH .
create-title-validate-length.json - изменен: title - константа длиной 50 символов, amount - 0.00.
create-title-invalidate-length.json - добавлен json-файл для тестирования длины поля title длиной 51.

AllaVrublevskaja committed:
AccountsComponentTest.java - изменены тесты на соответствие длины title схеме
AccountJson.java - добавлена константа CREATE_TITLE_INVALIDATE_LENGTH .
create-title-validate-length.json - изменен: title - константа длиной 50 символов, amount - 0.00.
create-title-invalidate-length.json - добавлен json-файл для тестирования длины поля title длиной 51.

AllaVrublevskaja committed:
openapi.yaml - удалила maxLength: 50 для AccountResponse title
